File: test/stat.rb
# MIT License
#
# Copyright (c) Sebastian Katzer 2017
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

assert 'SFTP::Stat' do
  assert_kind_of Class, SFTP::Stat
  assert_include SFTP::Stat.constants, :T_DIRECTORY
  assert_include SFTP::Stat.constants, :T_REGULAR
end

assert 'SFTP::Stat#initialize' do
  assert_nothing_raised { SFTP::Stat.new }
  assert_nothing_raised { SFTP::Stat.new uid: 1 }
  assert_equal SFTP::Stat::T_UNKNOWN, SFTP::Stat.new.type

  attrs = { gid: 1, uid: 2, mtime: 3, atime: 4, mode: 5, type: 6, size: 7 }
  stats = SFTP::Stat.new(attrs)

  assert_equal attrs[:gid], stats.gid
  assert_equal attrs[:uid], stats.uid
  assert_equal attrs[:mtime], stats.mtime
  assert_equal attrs[:atime], stats.atime
  assert_equal attrs[:mode], stats.mode
  assert_equal SFTP::Stat::T_UNKNOWN, stats.type
  assert_nil stats.size
end

assert 'SFTP::Stat#gid=' do
  stats = SFTP::Stat.new

  assert_nil stats.gid
  stats.gid = 0
  assert_equal 0, stats.gid
end

assert 'SFTP::Stat#uid=' do
  stats = SFTP::Stat.new

  assert_nil stats.uid
  stats.uid = 0
  assert_equal 0, stats.uid
end

assert 'SFTP::Stat#mtime=' do
  stats = SFTP::Stat.new

  assert_nil stats.mtime
  stats.mtime = 0
  assert_equal 0, stats.mtime
end

assert 'SFTP::Stat#atime=' do
  stats = SFTP::Stat.new

  assert_nil stats.atime
  stats.atime = 0
  assert_equal 0, stats.atime
end

assert 'SFTP::Stat#mode=' do
  stats = SFTP::Stat.new

  assert_nil stats.mode
  stats.mode = 0
  assert_equal 0, stats.mode
end

assert 'SFTP::Stat#size=' do
  stats = SFTP::Stat.new

  assert_nil stats.size
  assert_raise(NoMethodError) { stats.size = 1 }
  assert_nil stats.size
end

assert 'SFTP::Stat#type=' do
  stats = SFTP::Stat.new

  assert_equal SFTP::Stat::T_UNKNOWN, stats.type
  assert_raise(NoMethodError) { stats.type = SFTP::Stat::T_REGULAR }
  assert_equal SFTP::Stat::T_UNKNOWN, stats.type
end

assert 'SFTP::Stat#type' do
  stats = SFTP::Stat.new

  assert_equal SFTP::Stat::T_UNKNOWN, stats.type
  stats.mode = 0o140000
  assert_equal SFTP::Stat::T_SOCKET, stats.type
  stats.mode = 0o120000
  assert_equal SFTP::Stat::T_SYMLINK, stats.type
  stats.mode = 0o100000
  assert_equal SFTP::Stat::T_REGULAR, stats.type
  stats.mode = 0o60000
  assert_equal SFTP::Stat::T_BLOCK_DEVICE, stats.type
  stats.mode = 0o40000
  assert_equal SFTP::Stat::T_DIRECTORY, stats.type
  stats.mode = 0o20000
  assert_equal SFTP::Stat::T_CHAR_DEVICE, stats.type
  stats.mode = 0o10000
  assert_equal SFTP::Stat::T_FIFO, stats.type
end

assert 'SFTP::Stat#ftype' do
  stats = SFTP::Stat.new

  assert_equal :unknown, stats.ftype
  stats.mode = 0o140000
  assert_equal :socket, stats.ftype
  stats.mode = 0o120000
  assert_equal :symlink, stats.ftype
  stats.mode = 0o100000
  assert_equal :regular, stats.ftype
  stats.mode = 0o60000
  assert_equal :block_device, stats.ftype
  stats.mode = 0o40000
  assert_equal :directory, stats.ftype
  stats.mode = 0o20000
  assert_equal :char_device, stats.ftype
  stats.mode = 0o10000
  assert_equal :fifo, stats.ftype
end

assert 'SFTP::Stat#directory?' do
  stats = SFTP::Stat.new

  assert_nil stats.directory?
  stats.mode = 0o40000
  assert_true stats.directory?
  stats.mode = 0o10000
  assert_false stats.directory?
  stats.mode = 0o0
  assert_nil stats.directory?
end

assert 'SFTP::Stat#file?' do
  stats = SFTP::Stat.new

  assert_nil stats.file?
  stats.mode = 0o100000
  assert_true stats.file?
  stats.mode = 0o10000
  assert_false stats.file?
  stats.mode = 0o0
  assert_nil stats.file?
end

assert 'SFTP::Stat#symlink?' do
  stats = SFTP::Stat.new

  assert_nil stats.symlink?
  stats.mode = 0o120000
  assert_true stats.symlink?
  stats.mode = 0o10000
  assert_false stats.symlink?
  stats.mode = 0o0
  assert_nil stats.symlink?
end

assert 'SFTP::Stat#zero?' do
  stats = SFTP::Stat.new

  assert_true stats.zero?
  stats.instance_variable_set :@size, 240
  assert_false stats.zero?
  stats.instance_variable_set :@size, 0
  assert_true stats.zero?
end

assert 'SFTP::Stat#readable?' do
  stats = SFTP::Stat.new mode: SFTP::Stat::P_USR_X

  assert_false stats.readable?
  stats.mode = SFTP::Stat::P_USR_R
  assert_true stats.readable?
  stats.mode = SFTP::Stat::P_GRP_R
  assert_true stats.readable?
  stats.mode = SFTP::Stat::P_OTH_R
  assert_true stats.readable?
  stats.mode = SFTP::Stat::P_OTH_RWX
  assert_true stats.readable?
end

assert 'SFTP::Stat', 'SFTP' do
  SFTP.start('test.rebex.net', 'demo', password: 'password') do |sftp|
    dummy = sftp.file.stat('I am bad')
    assert_equal SFTP::Stat::T_UNKNOWN, dummy.type

    dir = sftp.file.stat('/pub')
    assert_true dir.directory?
    assert_equal 0, dir.size

    file = sftp.file.stat('readme.txt')
    assert_true file.file?
    assert_equal 600, file.umode
    assert_equal 403, file.size
    assert_true file.readable?
    assert_true file.writable?
    assert_false file.executable?
  end
end

File: src/stat.c
/* MIT License
 *
 * Copyright (c) Sebastian Katzer 2017
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "stat.h"

#include "mruby.h"
#include "mruby/hash.h"
#include "mruby/variable.h"

#include <libssh2_sftp.h>

#define SYM(name, len) mrb_intern_static(mrb, name, len)

mrb_value
mrb_sftp_stat_obj (mrb_state *mrb, LIBSSH2_SFTP_ATTRIBUTES *attrs)
{
    struct RClass *ftp, *cls;
    mrb_value obj;

    ftp = mrb_module_get(mrb, "SFTP");
    cls = mrb_class_get_under(mrb, ftp, "Stat");
    obj = mrb_obj_new(mrb, cls, 0, NULL);

    if (!attrs) return obj;

    if (attrs->flags & LIBSSH2_SFTP_ATTR_ACMODTIME) {
        mrb_iv_set(mrb, obj, SYM("@atime", 6), mrb_fixnum_value(attrs->atime));
        mrb_iv_set(mrb, obj, SYM("@mtime", 6), mrb_fixnum_value(attrs->mtime));
    }

    if (attrs->flags & LIBSSH2_SFTP_ATTR_SIZE) {
        mrb_iv_set(mrb, obj, SYM("@size", 5), mrb_fixnum_value(attrs->filesize));
    }

    if (attrs->flags & LIBSSH2_SFTP_ATTR_PERMISSIONS) {
        mrb_iv_set(mrb, obj, SYM("@mode", 5), mrb_fixnum_value(attrs->permissions));
    }

    if (attrs->flags & LIBSSH2_SFTP_ATTR_UIDGID) {
        mrb_iv_set(mrb, obj, SYM("@uid", 4), mrb_fixnum_value(attrs->uid));
        mrb_iv_set(mrb, obj, SYM("@gid", 4), mrb_fixnum_value(attrs->gid));
    }

    return obj;
}

void
mrb_sftp_hash_to_stat (mrb_state *mrb, mrb_value hsh, LIBSSH2_SFTP_ATTRIBUTES *attrs)
{
    mrb_value atime = mrb_hash_get(mrb, hsh, mrb_symbol_value(SYM("atime", 5)));
    mrb_value mtime = mrb_hash_get(mrb, hsh, mrb_symbol_value(SYM("mtime", 5)));
    mrb_value gid   = mrb_hash_get(mrb, hsh, mrb_symbol_value(SYM("gid", 3)));
    mrb_value uid   = mrb_hash_get(mrb, hsh, mrb_symbol_value(SYM("uid", 3)));
    mrb_value size  = mrb_hash_get(mrb, hsh, mrb_symbol_value(SYM("size", 4)));
    mrb_value mode  = mrb_hash_get(mrb, hsh, mrb_symbol_value(SYM("mode", 4)));

    attrs->flags = 0;

    if (mrb_test(atime) || mrb_test(mtime)) {
        attrs->flags |= LIBSSH2_SFTP_ATTR_ACMODTIME;
        attrs->atime = mrb_fixnum(atime);
        attrs->mtime = mrb_fixnum(mtime);
    }

    if (mrb_test(uid) || mrb_test(gid)) {
        attrs->flags |= LIBSSH2_SFTP_ATTR_UIDGID;
        attrs->uid = mrb_fixnum(uid);
        attrs->gid = mrb_fixnum(gid);
    }

    if (mrb_test(size)) {
        attrs->flags |= LIBSSH2_SFTP_ATTR_SIZE;
        attrs->filesize = mrb_fixnum(size);
    }

    if (mrb_test(mode)) {
        attrs->flags |= LIBSSH2_SFTP_ATTR_PERMISSIONS;
        attrs->permissions = mrb_fixnum(mode);
    }
}

static mrb_value
mrb_sftp_f_type (mrb_state *mrb, mrb_value self)
{
    mrb_value mode = mrb_attr_get(mrb, self, SYM("@mode", 5));
    unsigned int m = mrb_fixnum(mode);

    if (mrb_nil_p(mode) || m == 0)
        mrb_fixnum_value(LIBSSH2_SFTP_TYPE_UNKNOWN);

    if (LIBSSH2_SFTP_S_ISLNK(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_SYMLINK);

    if (LIBSSH2_SFTP_S_ISREG(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_REGULAR);

    if (LIBSSH2_SFTP_S_ISDIR(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_DIRECTORY);

    if (LIBSSH2_SFTP_S_ISCHR(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_CHAR_DEVICE);

    if (LIBSSH2_SFTP_S_ISBLK(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_BLOCK_DEVICE);

    if (LIBSSH2_SFTP_S_ISFIFO(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_FIFO);

    if (LIBSSH2_SFTP_S_ISSOCK(m))
        return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_SOCKET);

    return mrb_fixnum_value(LIBSSH2_SFTP_TYPE_UNKNOWN);
}

void
mrb_mruby_sftp_stat_init (mrb_state *mrb)
{
    struct RClass *ftp, *cls;

    ftp = mrb_module_get(mrb, "SFTP");
    cls = mrb_define_class_under(mrb, ftp, "Stat", mrb->object_class);

    mrb_define_method(mrb, cls, "type", mrb_sftp_f_type, MRB_ARGS_NONE());

    mrb_define_const(mrb, cls, "T_REGULAR", mrb_fixnum_value(LIBSSH2_SFTP_TYPE_REGULAR));
    mrb_define_const(mrb, cls, "T_DIRECTORY", mrb_fixnum_value(LIBSSH2_SFTP_TYPE_DIRECTORY));
    mrb_define_const(mrb, cls, "T_SYMLINK", mrb_fixnum_value(LIBSSH2_SFTP_TYPE_SYMLINK));
    mrb_define_const(mrb, cls, "T_UNKNOWN", mrb_fixnum_value(LIBSSH2_SFTP_TYPE_UNKNOWN));
    mrb_define_const(mrb, cls, "T_SOCKET",  mrb_fixnum_value(LIBSSH2_SFTP_TYPE_SOCKET));
    mrb_define_const(mrb, cls, "T_CHAR_DEVICE", mrb_fixnum_value(LIBSSH2_SFTP_TYPE_CHAR_DEVICE));
    mrb_define_const(mrb, cls, "T_BLOCK_DEVICE", mrb_fixnum_value(LIBSSH2_SFTP_TYPE_BLOCK_DEVICE));
    mrb_define_const(mrb, cls, "T_FIFO",    mrb_fixnum_value(LIBSSH2_SFTP_TYPE_FIFO));

    mrb_define_const(mrb, cls, "P_USR_RWX", mrb_fixnum_value(LIBSSH2_SFTP_S_IRWXU));
    mrb_define_const(mrb, cls, "P_USR_R",   mrb_fixnum_value(LIBSSH2_SFTP_S_IRUSR));
    mrb_define_const(mrb, cls, "P_USR_W",   mrb_fixnum_value(LIBSSH2_SFTP_S_IWUSR));
    mrb_define_const(mrb, cls, "P_USR_X",   mrb_fixnum_value(LIBSSH2_SFTP_S_IXUSR));

    mrb_define_const(mrb, cls, "P_GRP_RWX", mrb_fixnum_value(LIBSSH2_SFTP_S_IRWXG));
    mrb_define_const(mrb, cls, "P_GRP_R",   mrb_fixnum_value(LIBSSH2_SFTP_S_IRGRP));
    mrb_define_const(mrb, cls, "P_GRP_W",   mrb_fixnum_value(LIBSSH2_SFTP_S_IWGRP));
    mrb_define_const(mrb, cls, "P_GRP_X",   mrb_fixnum_value(LIBSSH2_SFTP_S_IXGRP));

    mrb_define_const(mrb, cls, "P_OTH_RWX", mrb_fixnum_value(LIBSSH2_SFTP_S_IRWXO));
    mrb_define_const(mrb, cls, "P_OTH_R",   mrb_fixnum_value(LIBSSH2_SFTP_S_IROTH));
    mrb_define_const(mrb, cls, "P_OTH_W",   mrb_fixnum_value(LIBSSH2_SFTP_S_IWOTH));
    mrb_define_const(mrb, cls, "P_OTH_X",   mrb_fixnum_value(LIBSSH2_SFTP_S_IXOTH));
}
